pkg/tso/keyspace_group_manager.go

```diff
@@ -146,6 +153,43 @@ func (s *state) getDeletedGroupNum() int {
 	return len(s.deletedGroups)
 }
 
+// cleanKeyspaceGroup cleans the given keyspace group from the state.
+// NOTICE: currently the only legal way to delete a keyspace group is
+// to merge it into another one. This function is used to clean up the
+// remaining info after the merge has been finished.
+func (s *state) cleanKeyspaceGroup(groupID uint32) {
+	s.Lock()
+	defer s.Unlock()
+	delete(s.deletedGroups, groupID)
+	delete(s.requestedGroups, groupID)
+}
+
+// markGroupRequested checks if the given keyspace group has been requested and should be marked.
+// If yes, it will do nothing and return nil directly.
+// If not, it will try to mark the keyspace group as requested inside a critical section, which
+// will call the checker passed in to check if the keyspace group is qualified to be marked as requested.
+// Any error encountered during the check will be returned to the caller.
+func (s *state) markGroupRequested(groupID uint32, checker func() error) error {
+	// Fast path to check if the keyspace group has been marked as requested.
+	s.RLock()
+	_, ok := s.requestedGroups[groupID]
+	s.RUnlock()
+	if ok {
+		return nil
+	}
+	s.Lock()
+	defer s.Unlock()
+	// Double check if the keyspace group has been marked as requested.
+	if _, ok := s.requestedGroups[groupID]; ok {
```

Comment:
Is it necessary to double check it?

Reply:
I think this may happen in the case of high concurrency. Double checking can reduce the possibility of etcd transaction conflicts in sync timestamp.

Reply:
Just to avoid the unnecessary operation as much as possible on the TSO hot path.